lib/common/xml.c

```diff
@@ -1860,79 +1997,32 @@ replace_node(xmlNode *old, xmlNode *new)
  * \return Standard Pacemaker return code
  */
 int
-pcmk__xe_find_replace(xmlNode *parent, xmlNode *child, xmlNode *update,
-                      bool delete_only)
+pcmk__xe_find_replace(xmlNode *xml, xmlNode *update, bool delete_only)
 {
-    /* @TODO Create a recursive helper so that callers don't need to pass a NULL
-     * parent argument.
-     *
-     * @TODO Try to extract delete into a different function from replace.
-     * Ideally separate tree traversal from any work.
-     *
-     * @COMPAT Is pcmk__apply_acl() enough to properly consider ACLs?
+    /* @COMPAT Is pcmk__apply_acl() enough to properly consider ACLs?
```

Comment:
Well, the delete function directly checks ACLs: if it finds a match to delete, it calls `free_xml()`, which calls `pcmk__check_acl()`.

The replace function basically just does this:
```
    new = xmlCopyNode(new, 1);
    pcmk__mem_assert(new);

    // May be unnecessary but avoids slight changes to some test outputs
    reset_xml_node_flags(new);

    old = xmlReplaceNode(old, new);

    if (xml_tracking_changes(new)) {
        // Replaced sections may have included relevant ACLs
        pcmk__apply_acl(new);
    }
    xml_calculate_changes(old, new);
    xmlFreeNode(old);
```

In either case, the old XML is freed, but in the replace case, that deletion seems to happen before ACLs are checked.

I've never enjoyed the change tracking and ACL code, and I have a tenuous grasp on it as of right now... maybe the `xml_calculate_changes()` call allows for reverting to `old` by undoing the changes, even if we've freed it.

Reply:
> `pcmk__xml_update()`

The `as_diff == true` case is actually deprecated. An upcoming PR adds a COMPAT note.

Reply:
ACLs don't prevent XML from being modified, they prevent it from being accepted as the new CIB. I believe it's the `xml_acl_denied()` call in `cib_perform_op()`.

`xml_calculate_changes()` compares the old and new XML, so it shouldn't matter what changes were made. The new XML should be marked appropriately.

But I forget how the ACL code works every time I deal with it, so that may be wrong.

Reply:
ACLs are only relevant for the CIB, so other XML (diffs, API requests, etc.) doesn't need to handle them

Reply:
Ah okay, I figured destroying part of the XML without checking ACLs first would be a problem, but it sounds like that's fine and we're assuming we're working with a scratch copy.